students/GKim/lesson07/html_render.py

```diff
@@ -0,0 +1,302 @@
+#!/usr/bin/env python3
+
+"""
+A class-based system for rendering html.
+This is someone elses code...very difficult assignment and I did
+I did not learn from it
+"""
+
+
+# This is the framework for the base class
+class Element():
+    """
+    Framework for a basic element in HTML code.
+    Class attributes:
+    tag:     Tag used for rendering html contents.  Element class is set to html.
+    indent:  The level of indentation for the current element.
+    """
+    tag = 'html'
+    indent = ''
+
+    def __init__(self, content=None, **kwargs):
+        """
+        __init__ method for Element class.
+        :param content: Content that is to be added to the instance content and later rendered.
+        :param kwargs:  Used for passing in tag attributes.
+        Attributes:
+        contents    The content that is to be rendered into HTML format.
+        """
+        self.attributes = kwargs
+        self.contents = [content] if content else []
+
+    def _open_tag(self):
+        """ Returns the opening tag for the current element """
+        return f'<{self.tag}>'
+
+    def _close_tag(self):
+        """ Returns the closing tag for the current element """
+        return f'</{self.tag}>'
+
+    def append(self, new_content):
+        """
+        Appends new_content to the instance attribute content
+        :param new_content: String that is to be appended to the instance content.
+        """
+        self.contents.append(new_content)
+
+    def render(self, out_file, indent="", ind_count=0):
+        """
+        Recursively renders the instance attribute content into HTML format code.
+        :param out_file: Destination for rendered text.
+        :param indent: The specified indentation level for elements.
+        :param ind_count: How many times indent should be applied to the element.
+        """
+        self.indent = indent * ind_count
+
+        out_file.write(f'{self.indent}{self._open_tag()[:-1]}')
+        for key, value in self.attributes.items():
+            out_file.write(f' {key}="{value}"')
+        out_file.write(f'{self._open_tag()[-1:]}\n')
+
+        for content in self.contents:
+            if hasattr(content, 'render'):
+                content.render(out_file, indent, ind_count + 1)
+            else:
+                out_file.write(f'{self.indent}{indent}{content}\n')
+
+        out_file.write(f'{self.indent}{self._close_tag()}\n')
```

Comment:
and then:

out_file.write(f'{cur_indent}{self._close_tag()}\n')

Key here is to think about what each of these names represetn, so you can know where to keep them:

"self" is for keeping teh current state of object, so that multiple methods all have access. So ony store things there that will be needed by other methods.

in this case:

self.indent is a class attribute, that represents a single level of indentation for ALL Elements -- youdo'nt want to change that.

ind_count is a function local variable that indicates how many levels of indentation this particular element needs to be indented while it is being rendered.

cur_indent (= self.indent * ind_count)is the current level of indentation for this element, while it is being rndered -- so it is only needed inside the render function, so no need to assign to self.

-CHB